### ISSUE TITLE
[clang-tidy]: `bugprone-unchecked-optional-access` false-positive if inside a loop we use checked optional value (might be a regression of CFG)

### ISSUE THREAD

**Issue by BaLiKfromUA** [llvm/llvm-project#174250]
## Way to reproduce

Smallest example which me and my colleague managed to make:
```cpp
#include <optional>
#include <vector>

std::vector<std::optional<int>> getOptionals();

void test() {

  for (const auto &opt : getOptionals()) {
      if (opt.has_value()) {
        int x = opt.value(); // NO ERROR
      }
  }

  std::optional<int> opt; 
  if (opt.has_value()) {
    for (int i = 0; i < 10; ++i) {
        int x = opt.value(); // NO ERROR
    }
  }  

  for (const auto &opt : getOptionals()) {
      if (opt.has_value()) {
          for (int i = 0; i < 10; ++i) {
            int x = opt.value(); // ERROR
          }
      }
  }

  // Also happens when the loop is unbounded
  for (const auto &opt : getOptionals()) {
      if (opt.has_value()) {
          for (;;) {
            int x = opt.value(); // ERROR
          }
      }
  }
}
```

[Compiler Explorer Link](https://compiler-explorer.com/z/9e6zjor5T)

## Initial problem / User Impact

After fixing https://github.com/llvm/llvm-project/pull/168863 I found new false-positive like:
```cpp
bsl::vector<bsl::optional<bsl::string>> getOptionals();

void test_with_ball_log_and_loop() {
    BALL_LOG_SET_CATEGORY(__func__);
  
    for (const auto& opt : getOptionals()) {
        if (opt.has_value()) {
            BALL_LOG_INFO << opt.value(); // ERROR?!
        }
    }
}
```

`BALL_LOG` is BDE tooling used for logging. [Link](https://bloomberg.github.io/bde-resources/doxygen/bde_api_prod/group__ball__log.html)

After some experiments, I found out that I can reproduce the same issue with `std::optional<int>` so it was somehow connected with `BALL_LOG`.

[Compiler Explorer Link](https://compiler-explorer.com/z/vdTvqrdv1).

Removing/expanding macros and includes resulted in example from the beginning :) 

---

**So even though the initial example might be niche, the original issue is pretty common pattern reported by users.**

JFYI, I cannot reproduce it with clang-18 and clang-19 releases but I can reproduce it with clang-20 and clang-21. So maybe it's a regression??

**Comment by BaLiKfromUA** [comment]
I am happy to contribute the fix and will start looking for a root cause soon.

My suspicion is that it is connected to Dataflow Framework, not the Model itself.

Since my familiarity with that part of code is almost zero, I would appreciate initial guidance and help :)

Thanks!

**Comment by BaLiKfromUA** [comment]
Just as another experiment, false-positive doesn't happen for code like this:
```cpp
    for (const auto &opt : getOptionals()) {
      if (opt.has_value()) {
          int x = opt.value(); // NO ERROR
          {
            int x = opt.value(); // NO ERROR
          }
      }
  }
```

So it might be related with looping inside loop, not adding new scope block

**Comment by BaLiKfromUA** [comment]
@zeyi2  Thank you a lot! These are very useful findings! 

I am more familiar with Model codebase so I might be able to find a fix.

I am now travelling so will try to look on Sunday or next Monday for a fix.

> IMO there could be a flaw in the valueCall matcher (maybe related to https://github.com/llvm/llvm-project/commit/58fe7f9683a020a880426a4d8d61b067b83a9380 or https://github.com/llvm/llvm-project/commit/2f0630f8bc91e81fd5948e82164cf82ae595caf2). In range-based for loops with reference types like const auto &opt fails to capture the call.

https://github.com/llvm/llvm-project/commit/58fe7f9683a020a880426a4d8d61b067b83a9380 looks for me more related then https://github.com/llvm/llvm-project/commit/6761b24ae2f34b923df46412475a9ece50542b97

How https://github.com/llvm/llvm-project/commit/2f0630f8bc91e81fd5948e82164cf82ae595caf2 I don't understand yet since it just extends list of checked types, no?

**Comment by zeyi2** [comment]
> How [2f0630f](https://github.com/llvm/llvm-project/commit/2f0630f8bc91e81fd5948e82164cf82ae595caf2) I don't understand yet since it just extends list of checked types, no?

Ah thanks for pointing that out, I just ran git blame and didn't do a more careful check.
